text-2.0 candidate release

I believe current `master` is mature enough to publish a candidate release for user acceptance testing. If there are no other suggestions, I'll upload it on Friday.

On GHA I got an error with GHC-8.2 and GHC-8.0

```
[1 of 1] Compiling Main             ( test/test.hs, /__w/scientific/scientific/dist-newstyle/build/x86_64-linux/ghc-8.0.2/scientific-0.3.7.0/t/test-scientific/build/test-scientific/test-scientific-tmp/Main.o )
Linking /__w/scientific/scientific/dist-newstyle/build/x86_64-linux/ghc-8.0.2/scientific-0.3.7.0/t/test-scientific/build/test-scientific/test-scientific ...
/github/home/.cabal/store/ghc-8.0.2/text-2.0-c46353126c38c0002af2c33e37c7db2e3337f6f1aea8545686900c7ce3780a4e/lib/libHStext-2.0-c46353126c38c0002af2c33e37c7db2e3337f6f1aea8545686900c7ce3780a4e.a(measure_off.o): In function `has_avx512_vl_bw':
measure_off.c:(.text+0x2df): undefined reference to `__get_cpuid_count'
```

https://github.com/basvandijk/scientific/runs/4674626374?check_suite_focus=true

I was able to reproduce in a xenial Docker container following that build job:

```
% docker run -ti --rm buildpack-deps:xenial

apt-get update
apt-get install -y --no-install-recommends gnupg ca-certificates dirmngr curl git software-properties-common
apt-add-repository -y 'ppa:hvr/ghc'
apt-get update
apt-get install -y ghc-8.0.2 cabal-install-3.4
export PATH=/opt/ghc/bin:$PATH

git clone https://github.com/basvandijk/scientific.git
cd scientific

curl -L https://github.com/basvandijk/scientific/pull/84.patch | git am

cabal update
cabal v2-build --disable-tests --disable-benchmarks --dependencies-only -j2
cabal v2-build --disable-tests --disable-benchmarks all
cabal v2-build --enable-tests --disable-benchmarks all
```

```
Linking C:\WORK\dist-newstyle\build\x86_64-windows\ghc-8.2.2\postgresql-simple-0.6.4\t\test\build\test\test.exe ...
C:\SR\ghc-8.2.2\text-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb\lib/libHStext-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb.a(measure_off.o):measure_off.c:(.text+0x2da): undefined reference to `__get_cpuid_count'
```

issue happened on oldish Windows used by appveyor: https://ci.appveyor.com/project/phadej/postgresql-simple/builds/42055973/job/66uasdf3b0il1k2j

I hope #401 will fix it.

I ran into a similar issue as https://github.com/haskell/text/issues/385#issuecomment-1003603159 when building `text-show` with `text-2.0` on Windows with GHC 8.2.2 ([CI log](https://ci.appveyor.com/project/RyanGlScott/text-show/builds/42072195/job/a9nh6qrtdpcq7rsc)):

```
[ 8 of 66] Compiling TextShow.GHC.Conc.Windows ( src\TextShow\GHC\Conc\Windows.hs, C:\WORK\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-show-3.9.5\build\TextShow\GHC\Conc\Windows.o )
ghc.exe:  | C:\SR\ghc-8.2.2\text-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb\lib\libHStext-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb.a: unknown symbol `__get_cpuid_count'
ghc.exe:  | C:\SR\ghc-8.2.2\text-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb\lib\libHStext-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb.a: unknown symbol `_hs_text_measure_off'
ghc.exe:  | C:\SR\ghc-8.2.2\text-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb\lib\libHStext-2.0-5e3f12ca9a37edf5a48459f7f3969564285f2adb.a: unknown symbol `textzm2zi0zm5e3f12ca9a37edf5a48459f7f3969564285f2adb_DataziTextziLazzy_zdfShowText2_closure'
ghc.exe:  | C:\WORK\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-show-3.9.5\build\TextShow\Classes.o: unknown symbol `textzm2zi0zm5e3f12ca9a37edf5a48459f7f3969564285f2adb_DataziTextziInternalziBuilder_Buffer_con_info'
```

I tried building the `master` branch of `text` as well, but that failed in a similar fashion:

```
$ cabal build -w ../ghc-8.2.2/bin/ghc
Build profile: -w ghc-8.2.2 -O1
In order, the following will be built (use -v for more details):
 - text-2.0 (test:tests) (first run)
 - text-2.0 (bench:text-benchmarks) (first run)
Preprocessing benchmark 'text-benchmarks' for text-2.0..
Preprocessing test suite 'tests' for text-2.0..
Building test suite 'tests' for text-2.0..
Building benchmark 'text-benchmarks' for text-2.0..
[ 1 of 16] Compiling Tests.Lift       ( tests\Tests\Lift.hs, C:\Users\winferno\Software\text\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-2.0\t\tests\build\tests\tests-tmp\Tests\Lift.o )
Linking C:\Users\winferno\Software\text\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-2.0\b\text-benchmarks\build\text-benchmarks\text-benchmarks.exe ...
ghc.exe:  | C:\Users\winferno\Software\text\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-2.0\build\libHStext-2.0-inplace.a: unknown symbol `__get_cpuid_count'
ghc.exe:  | C:\Users\winferno\Software\text\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-2.0\build\libHStext-2.0-inplace.a: unknown symbol `_hs_text_measure_off'
ghc.exe: ^^ Could not load 'textzm2zi0zminplace_DataziText_zdfIsStringText_closure', dependency unresolved. See top entry above.


ByteCodeLink.lookupCE
During interactive linking, GHCi couldn't find the following symbol:
  textzm2zi0zminplace_DataziText_zdfIsStringText_closure
This may be due to you not asking GHCi to load extra object files,
archives or DLLs needed by your current session.  Restart GHCi, specifying
the missing library using the -L/path/to/object/dir and -lmissinglibname
flags, or simply by naming the relevant files on the GHCi command line.
Alternatively, this link failure might indicate a bug in GHCi.
If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org

C:\Users\winferno\Software\text\dist-newstyle\build\x86_64-windows\ghc-8.2.2\text-2.0\build/libHStext-2.0-inplace.a(measure_off.o):measure_off.c:(.text+0x2da): undefined reference to `__get_cpuid_count'
collect2.exe: error: ld returned 1 exit status
`gcc.exe' failed in phase `Linker'. (Exit code: 1)
cabal.exe: Failed to build test:tests from text-2.0.
Failed to build bench:text-benchmarks from text-2.0.
```

I'm unclear if anything can be done about this on `text`'s end, but I wanted to mention it regardless. Let me know if this should be a separate issue.

--- a/text.cabal
+++ b/text.cabal
@@ -99,11 +99,11 @@ library
 
   -- Certain version of GHC crash on Windows, when TemplateHaskell encounters C++.
   -- https://gitlab.haskell.org/ghc/ghc/-/issues/19417
-  if flag(simdutf) && os(windows) && (impl(ghc == 8.0.1) || impl(ghc >= 8.8) && impl(ghc < 8.10.5) || impl(ghc == 9.0.1))
+  if flag(simdutf) && os(windows) && impl(ghc == 8.0.1 || >= 8.8 && < 8.10.5 || == 9.0.1)
     build-depends: base < 0
 
   -- For GHC 8.2, 8.6.3 and 8.10.1 even TH + C crash Windows linker.
-  if os(windows) && (impl(ghc >= 8.2) && impl(ghc < 8.4) || impl(ghc == 8.6.3) || impl(ghc == 8.10.1))
+  if os(windows) && impl(ghc >= 8.2 && < 8.4 || == 8.6.3 || == 8.10.1)
     build-depends: base < 0
 
   exposed-modules:
@@ -239,7 +239,7 @@ test-suite tests
   -- Starting from 9.2 ghc library depends on parsec,
   -- which causes circular dependency.
   -- Plugin infrastructure does not work properly in 8.0.1 and 8.6.1
-  if impl(ghc >= 8.0.2) && impl(ghc < 8.6) || impl(ghc >= 8.6.2) && impl(ghc < 9.2)
+  if impl(ghc >= 8.0.2 && < 8.6 || >= 8.6.2 && < 9.2)
     build-depends: tasty-inspection-testing
 
   default-language: Haskell2010

--- a/text.cabal
+++ b/text.cabal
@@ -97,6 +97,15 @@ library
     else
       extra-libraries: stdc++
 
+  -- Certain version of GHC crash on Windows, when TemplateHaskell encounters C++.
+  -- https://gitlab.haskell.org/ghc/ghc/-/issues/19417
+  if flag(simdutf) && os(windows) && (impl(ghc == 8.0.1) || impl(ghc >= 8.8) && impl(ghc < 8.10.5) || impl(ghc == 9.0.1))
+    build-depends: base < 0
+
+  -- For GHC 8.2, 8.6.3 and 8.10.1 even TH + C crash Windows linker.
+  if os(windows) && (impl(ghc >= 8.2) && impl(ghc < 8.4) || impl(ghc == 8.6.3) || impl(ghc == 8.10.1))
+    build-depends: base < 0
+
   exposed-modules:
     Data.Text
     Data.Text.Array
@@ -229,7 +238,8 @@ test-suite tests
 
   -- Starting from 9.2 ghc library depends on parsec,
   -- which causes circular dependency.
-  if impl(ghc < 9.2)
+  -- Plugin infrastructure does not work properly in 8.0.1 and 8.6.1
+  if impl(ghc >= 8.0.2) && impl(ghc < 8.6) || impl(ghc >= 8.6.2) && impl(ghc < 9.2)
     build-depends: tasty-inspection-testing
 
   default-language: Haskell2010

